### ISSUE TITLE
Internal error in `Agda/TypeChecking/Monad/Context.hs` using Mimer

### ISSUE THREAD

**Issue by oskeri** [agda/agda#7639]
The following code triggers an internal error when trying to use Mimer (`C-c C-a`) in either of the last two holes:

```Agda
open import Agda.Builtin.Nat
open import Agda.Builtin.Sigma

data ⊥ : Set where

_×_ : (A B : Set) → Set
A × B = Σ A (λ _ → B)

impossible : Nat → Nat → ⊥ × ⊥
impossible m zero = {!!}
impossible m (suc n) =
  let _ , _ = impossible {!!} n
  in  {!!} , {!!}
```

This is the error:

```
An internal error has occurred. Please report this as a bug.
Location of the error: __IMPOSSIBLE__, called at
src/full\Agda\TypeChecking\Monad\Context.hs:139:32 in
Agda-2.8.0-9u1uvJH0I89CZ4mhZl7o05:Agda.TypeChecking.Monad.Context
```

The error happens both on version 2.7.0.1 and on latest master.

None of the following versions trigger the error:

```Agda
impossible₁ : Nat → Nat → ⊥
impossible₁ m zero = {!!}
impossible₁ m (suc n) =
  let _ = impossible₁ {!!} n
  in  {!!}

impossible₂ : Nat → ⊥ × ⊥
impossible₂ zero = {!!}
impossible₂ (suc n) =
  let _ , _ = impossible₂ n
  in  {!!} , {!!}

impossible₃ : Nat → Nat → ⊥ × ⊥
impossible₃ m zero = {!!}
impossible₃ m (suc n) =
  let _ = impossible₃ {!!} n
  in  {!!} , {!!}

impossible₄ : Nat → Nat → ⊥ × ⊥
impossible₄ m zero = {!!}
impossible₄ m (suc n) =
  let _ , _ = impossible₄ {!!} n
  in  {!!}
```

**Comment by andreasabel** [comment]
The error is raised when trying to find a checkpoint in the current `TCEnv`: https://github.com/agda/agda/blob/7b3b899815f9a23aeae874e9d45680a44f3a9b06/src/full/Agda/TypeChecking/Monad/Context.hs#L137-L139

By turning on debug printing, we traced it back to this line:
https://github.com/agda/agda/blob/7b3b899815f9a23aeae874e9d45680a44f3a9b06/src/full/Agda/Mimer/Mimer.hs#L895

Mimer is playing with checkpoints a lot and `let` (via `Open`) is also implemented with checkpoints, so it could well be that Mimer shoots itself in the foot somewhere.

Note that checkpoints are both handled
1. in the `TCState` (and state of Mimer); those persist across subcomputations
2. and in the `TCEnv` which does not persist across subcomputations

so it seems, from the bird's eye, a priori plausible they could get out of sync.

Since checkpoints entered the core of the type checker at some point, I would love to see a typing-rules presentation of this aspect so one could build some mental model and study the correctness conditions. @UlfNorell @jespercockx

**Comment by UlfNorell** [comment]
The mental model is quite simple: as the current context gets modified (for instance by going under a binder) you need to apply a substitution (a weakening in the case of going under a binder) to terms that were defined in the parent context if you want to use them in the new context. Checkpoints track this substitution, with `checkpointSubstitution` computing the substitution you need to apply to terms defined at the given checkpoint to use them in the current context.

The impossible is thrown if you ask for the substitution for a checkpoint that does not refer to a parent context of the current context.

**Comment by andreasabel** [comment]
Yes, but if you allocate a `CheckpointId` in a child context and then go back to the parent context the checkpoint goes out of scope but there are fields in the state that hold on to checkpoints, so they might store a checkpoint that is no longer in the `checkpointSubstitution` map.  And then asking for the substitution for this checkpoint just crashes.
If that is correct, we do not have a pretext for such IDs.  Our other IDs are either refering to the state (which persists), or they point into the context (de Bruijn indicies), but then they are abstracted by term formers (`Abs`).

**Comment by UlfNorell** [comment]
I'm not sure what you are trying to say.

Asking for the substitution for a checkpoint that is not from a parent context is wrong and should crash. It's like referencing a deBruijn index that's too big.

I believe the checkpoints stored in the state are only in the state for technical reasons to do with how modules get checked, they should behave as if everything was in the environment.

**Comment by andreasabel** [comment]
Judgements and rules.
Thanks.

**Comment by UlfNorell** [comment]
For what exactly? How to inline a let-binding from a different context? The checkpoint machinery doesn't add anything to the theory, it's a convenient way to remember which context a term comes from and how to bring it into the current context should you need to.

**Comment by jespercockx** [comment]
As Ulf said, checkpoints are more of an implementation technique than a modification to the type theory itself. Moreover, afaik they are used for elaboration (e.g. for getting the value of a let-bound variable after some context changes) but not for the core type theory itself (as evidenced by the fact that checkpoints are not used in `checkInternal`).

So technically one could imagine having formal *elaboration* judgements and rules that explicitly keep track of and use checkpoints. However Agda does not have formal elaboration rules for most of its features (our paper on elaborating dependent pattern matching being the major exception). So while it would be nice to have these formal elaboration rules, demanding them really seems like it is holding checkpoints to a higher standard than almost any other part of Agda's elaborator.

**Comment by andreasabel** [comment]
Well, we do have formal investigations into some aspects of Agda that go beyond the bare typing rules:
- meta-variables and constraints
- lhs unification and coverage
- termination
- polarities

This is actually quite a bit.
The papers usually investigate just a core of Agda, and need extrapolation, but still, they give some guidance and explanation that is independent of the Haskell code base of Agda.

Doing these formal investigations is an indispensable part of developing a proof assistant.

You might disagree, but then I am asking why one would
1. on the one side want to implement Martin-Löf Type Theory
2. but on the other side reject its methodology: judgments and rules.

To me, this is like writing a book on LaTeX in Microsoft Word.

**Comment by UlfNorell** [comment]
Ok, but I'm still not sure what rules you want to see. It wouldn't be impossible to write down some rules for what happens when we encounter a let-binding, how checkpoint substitutions are updated when extending the context and how they are applied when inlining a let-binding. I'm just not sure that this would give you any deep insights, since it's pretty straightforward.

**Comment by jespercockx** [comment]
Minimized test case:
```agda
record R : Set₁ where
  constructor c
  field f : Set

impossible : R
impossible = let c _ = c _ in  {!!} -- C-c C-a
```

**Comment by jespercockx** [comment]
After familiarizing myself a bit with the Mimer implementation I tracked down the problem to the implementation of `checkSolved`:

https://github.com/agda/agda/blob/50c9f1ff04aeafcc7548a4dad06e1b4a723a5cb2/src/full/Agda/Mimer/Mimer.hs#L1426-L1429

Here we set the list of new goals to *all* unsolved metas in the refinement of the goal we just found. This includes also metavariables that were already existing and created outside of the scope of the interaction hole we're trying to solve. This leads us to try to solve these old metas, but we are still using the same state which includes the local context and let bindings of the interaction hole. Then when we try to use one of these to solve the old meta, the checkpoint machinery rightfully tells us that we're not in the right scope to do so.

My first attempt to fix this was to simply not update the list of new goals; after all these are already set by other functions such as `tryRefineWith` (which *does* correctly add only new metas). However, this leads to some kind of looping behaviour as I'm getting a timeout in many test cases. 

The more conservative thing to do would be to update the list, but filter out any "old" metas (i.e. ones that do not have the current checkpoint in their history). I guess this would fix the problem, but I don't understand yet why updating the goal list is necessary at all.

**Comment by andreasabel** [comment]
Minor point of observation: I realized when I did some debug printing that `sbGoals` often contains the same meta several times.

**Comment by jespercockx** [comment]
> Minor point of observation: I realized when I did some debug printing that sbGoals often contains the same meta several times.

This is perhaps because of the use of the list monoid for the call to `allMetas` in the code I quoted above. Perhaps it would make more sense to use a set rather than a list (if the call to `allMetas` is even necessary at all).

**Comment by jespercockx** [comment]
Now that this issue is fixed and the call to allMetas is gone, could you check if you still get duplicate metas? I'd expect them to be gone now, but if they're still there we could consider switching to a set.
